File: packages/lhc/src/shared-tech/content-blocks.ts
// Anthropic Messages API content blocks inside LHC. Every block type the API
// defines is held faithfully; the only rewrite is that binary or opaque
// payload strings (base64 image and PDF bytes, redacted-thinking data, web
// search encrypted content) leave the JSON and live in the thread's blob table,
// keyed by content hash. The block keeps the API's shape and names; the payload
// field holds a reference `{ $blob, bytes }` in place of the string. Text never
// carries base64: not in the event payload, not in a message block, not in a
// band, not in a served view unless the caller asks for the block back inlined.
//
// Pure module: no database. Callers extract blobs at intake and inline them at
// serving through the two small functions here.
import { createHash } from "node:crypto";

/** Block type names, as the API spells them. Text-shaped types are stored as
 *  JSON verbatim (no blob); blob-bearing types have their payload paths
 *  extracted. */
export const API_BLOCK_TYPES = [
  "text",
  "image",
  "document",
  "tool_use",
  "tool_result",
  "thinking",
  "redacted_thinking",
  "server_tool_use",
  "web_search_tool_result",
  "web_fetch_tool_result",
  "code_execution_tool_result",
  "bash_code_execution_tool_result",
  "text_editor_code_execution_tool_result",
  "tool_search_tool_result",
  "search_result",
  "container_upload",
  "tool_reference",
] as const;
export type ApiBlockType = (typeof API_BLOCK_TYPES)[number];

export type ApiBlock = Record<string, unknown> & { type: string };

export interface BlobRef {
  $blob: string; // "sha256:<hex>"
  bytes: number; // decoded byte length
}

export interface ExtractedBlob {
  sha256: string;
  mediaType: string | null;
  data: Uint8Array;
}

export function isPlainRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function isApiBlock(value: unknown): value is ApiBlock {
  return (
    isPlainRecord(value) &&
    typeof value["type"] === "string" &&
    (API_BLOCK_TYPES as readonly string[]).includes(value["type"])
  );
}

export function isBlobRef(value: unknown): value is BlobRef {
  return isPlainRecord(value) && typeof value["$blob"] === "string" && typeof value["bytes"] === "number";
}

function blobRefFor(data: Uint8Array): { ref: BlobRef; sha256: string } {
  const sha256 = createHash("sha256").update(data).digest("hex");
  return { ref: { $blob: `sha256:${sha256}`, bytes: data.byteLength }, sha256 };
}

// ── extraction (intake) ──────────────────────────────────────────

interface Sink {
  blobs: ExtractedBlob[];
}

/** A `{type:"base64", media_type, data}` source: data leaves as a blob. Other
 *  source types (url, text, content) are text-shaped or nested. */
function extractSource(source: unknown, sink: Sink): unknown {
  if (!isPlainRecord(source)) return source;
  if (source["type"] === "base64" && typeof source["data"] === "string") {
    const data = Buffer.from(source["data"], "base64");
    const { ref, sha256 } = blobRefFor(data);
    sink.blobs.push({
      sha256,
      mediaType: typeof source["media_type"] === "string" ? source["media_type"] : null,
      data,
    });
    return { ...source, data: ref };
  }
  if (source["type"] === "content" && Array.isArray(source["content"])) {
    return { ...source, content: source["content"].map((inner) => extractBlock(inner, sink)) };
  }
  return source;
}

function extractOpaque(block: Record<string, unknown>, key: string, sink: Sink): Record<string, unknown> {
  const value = block[key];
  if (typeof value !== "string") return block;
  const data = Buffer.from(value, "utf8");
  const { ref, sha256 } = blobRefFor(data);
  sink.blobs.push({ sha256, mediaType: null, data });
  return { ...block, [key]: ref };
}

/** One block, blob payloads extracted, everything else verbatim. Unknown
 *  shapes pass through untouched (a newer API type is still a record). */
function extractBlock(value: unknown, sink: Sink): unknown {
  if (!isPlainRecord(value)) return value;
  switch (value["type"]) {
    case "image":
    case "document":
      return { ...value, source: extractSource(value["source"], sink) };
    case "redacted_thinking":
      return extractOpaque(value, "data", sink);
    case "tool_result":
      return Array.isArray(value["content"])
        ? { ...value, content: value["content"].map((inner) => extractBlock(inner, sink)) }
        : value;
    case "web_search_tool_result":
      return Array.isArray(value["content"])
        ? {
            ...value,
            content: value["content"].map((inner) =>
              isPlainRecord(inner) && inner["type"] === "web_search_result"
                ? extractOpaque(inner, "encrypted_content", sink)
                : inner,
            ),
          }
        : value;
    case "web_fetch_tool_result":
      // content is a web_fetch block whose `content` is a document block.
      return isPlainRecord(value["content"]) && isPlainRecord(value["content"]["content"])
        ? { ...value, content: { ...value["content"], content: extractBlock(value["content"]["content"], sink) } }
        : value;
    default:
      return value;
  }
}

export interface ExtractResult {
  blocks: ApiBlock[];
  blobs: ExtractedBlob[];
}

/** Blocks with their binary/opaque payloads replaced by blob references. */
export function extractBlobs(blocks: readonly unknown[]): ExtractResult {
  const sink: Sink = { blobs: [] };
  const out = blocks.map((block) => extractBlock(block, sink) as ApiBlock);
  return { blocks: out, blobs: sink.blobs };
}

// ── inlining (serving) ───────────────────────────────────────────

export type BlobLoader = (sha256: string) => Uint8Array | undefined;

function inlineValue(value: unknown, load: BlobLoader, encoding: "base64" | "utf8"): unknown {
  if (isBlobRef(value)) {
    const data = load(value.$blob.replace(/^sha256:/, ""));
    if (data === undefined) return value;
    return Buffer.from(data).toString(encoding);
  }
  return value;
}

/** The block as the API shaped it, blob payloads back in place. A blob that
 *  is missing from the store leaves its reference in place rather than
 *  inventing bytes. */
export function inlineBlobs(block: unknown, load: BlobLoader): unknown {
  if (Array.isArray(block)) return block.map((inner) => inlineBlobs(inner, load));
  if (!isPlainRecord(block)) return block;
  const out: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(block)) {
    if (isBlobRef(value)) {
      // base64 sources come back as base64; opaque strings (redacted thinking
      // data, encrypted search content) were stored as their utf8 bytes.
      const encoding = key === "data" && block["type"] === "base64" ? "base64" : "utf8";
      out[key] = inlineValue(value, load, encoding);
    } else if (isPlainRecord(value) || Array.isArray(value)) {
      out[key] = inlineBlobs(value, load);
    } else {
      out[key] = value;
    }
  }
  return out;
}

// ── text-shaped rendering (bands, derivations, retrieval) ────────

function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(0)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

function sourceDescription(source: unknown): string {
  if (!isPlainRecord(source)) return "";
  const mediaType = typeof source["media_type"] === "string" ? source["media_type"] : undefined;
  const data = source["data"];
  if (source["type"] === "base64" && isBlobRef(data))
    return [mediaType, formatBytes(data.bytes)].filter(Boolean).join(" · ");
  if (source["type"] === "base64" && typeof data === "string") {
    return [mediaType, formatBytes(Math.floor((data.length * 3) / 4))].filter(Boolean).join(" · ");
  }
  if (source["type"] === "url" && typeof source["url"] === "string") return source["url"];
  if (source["type"] === "text") return "text/plain";
  if (source["type"] === "content") return "content";
  return "";
}

/**
 * What the model needs to know that the block existed, not its content: the
 * API type, media type, size, and a title when the block has one. Text-shaped
 * blocks render their text. Tool-result nesting renders each inner block.
 */
export function placeholderText(block: unknown): string {
  if (!isPlainRecord(block)) return "";
  switch (block["type"]) {
    case "text":
      return typeof block["text"] === "string" ? block["text"] : "";
    case "image":
      return `[image · ${sourceDescription(block["source"])}]`;
    case "document": {
      const title = typeof block["title"] === "string" && block["title"] !== "" ? ` · ${block["title"]}` : "";
      const source = block["source"];
      if (isPlainRecord(source) && source["type"] === "text" && typeof source["data"] === "string")
        return source["data"];
      return `[document · ${sourceDescription(source)}${title}]`;
    }
    case "redacted_thinking":
      return "[redacted thinking]";
    case "search_result": {
      const title = typeof block["title"] === "string" ? block["title"] : "";
      const source = typeof block["source"] === "string" ? block["source"] : "";
      const inner = Array.isArray(block["content"]) ? block["content"].map(placeholderText).join("\n") : "";
      return `[search result · ${title} · ${source}]\n${inner}`;
    }
    case "tool_reference":
      return `[tool reference · ${String(block["tool_name"] ?? "")}]`;
    case "container_upload":
      return `[container upload · ${String(block["file_id"] ?? "")}]`;
    case "server_tool_use":
      return `[server tool use · ${String(block["name"] ?? "")}] ${JSON.stringify(block["input"] ?? {})}`;
    case "web_search_tool_result": {
      const content = block["content"];
      if (Array.isArray(content)) {
        const lines = content.map((r) =>
          isPlainRecord(r) ? `- ${String(r["title"] ?? "")} ${String(r["url"] ?? "")}`.trim() : "",
        );
        return `[web search result · ${content.length} result(s)]\n${lines.join("\n")}`;
      }
      return `[web search result] ${JSON.stringify(content ?? {})}`;
    }
    case "web_fetch_tool_result": {
      const content = block["content"];
      const url = isPlainRecord(content) && typeof content["url"] === "string" ? content["url"] : "";
      const doc = isPlainRecord(content) ? placeholderText(content["content"]) : "";
      return `[web fetch result · ${url}]\n${doc}`;
    }
    case "tool_result": {
      const content = block["content"];
      if (typeof content === "string") return content;
      return Array.isArray(content) ? content.map(placeholderText).join("\n") : "";
    }
    default:
      // code_execution_tool_result and friends are text-shaped JSON; anything
      // newer is at least made visible.
      return JSON.stringify(block);
  }
}

/** True when the block carries (or carried) a blob payload. */
export function hasBlobPayload(block: unknown): boolean {
  if (Array.isArray(block)) return block.some(hasBlobPayload);
  if (!isPlainRecord(block)) return false;
  return Object.values(block).some((value) => isBlobRef(value) || hasBlobPayload(value));
}

/**
 * Rough context cost of a block the text estimator cannot see. Images: the
 * API's ceiling for one image (~1,600 tokens at 1568px); PDFs: ~2,000 tokens
 * per page at ~50 KB a page, floor one page. Measured, not calibrated.
 */
export function blobTokenEstimate(block: unknown): number {
  if (!isPlainRecord(block)) return 0;
  const source = block["source"];
  const bytes = isPlainRecord(source) && isBlobRef(source["data"]) ? source["data"].bytes : 0;
  switch (block["type"]) {
    case "image":
      return 1_600;
    case "document":
      return bytes > 0 ? Math.max(1, Math.ceil(bytes / 50_000)) * 2_000 : 0;
    case "tool_result":
      return Array.isArray(block["content"])
        ? block["content"].reduce<number>((sum, inner) => sum + blobTokenEstimate(inner), 0)
        : 0;
    case "web_fetch_tool_result":
      return isPlainRecord(block["content"]) ? blobTokenEstimate(block["content"]["content"]) : 0;
    default:
      return 0;
  }
}

File: packages/lhc-rs/src/shared_tech/content_blocks.rs
//! Ported from packages/lhc/src/shared-tech/content-blocks.ts (TS `717a146`).
//!
//! Anthropic Messages API content blocks inside LHC. Every block type the API
//! defines is held faithfully; the only rewrite is that binary or opaque
//! payload strings (base64 image and PDF bytes, redacted-thinking data, web
//! search encrypted content) leave the JSON and live in the thread's blob
//! table, keyed by content hash. The block keeps the API's shape and names;
//! the payload field holds a reference `{ $blob, bytes }` in place of the
//! string. Text never carries base64: not in the event payload, not in a
//! message block, not in a band, not in a served view unless the caller asks
//! for the block back inlined.
//!
//! Pure module: no database. Callers extract blobs at intake and inline them
//! at serving through the two small functions here.
//!
//! Byte parity: rewritten blocks keep the source's insertion order with the
//! replaced key in place (JS object spread + same-key assignment; serde_json
//! `preserve_order` insert on an existing key keeps its slot). Hashes are
//! sha256 of the decoded bytes, matching Node `createHash("sha256")`.

use serde_json::{Map, Value};

use super::js_json::{js_json_stringify, js_string_nullish};
use super::sha256::sha256_hex_bytes;

/// Block type names, as the API spells them. Text-shaped types are stored as
/// JSON verbatim (no blob); blob-bearing types have their payload paths
/// extracted.
pub const API_BLOCK_TYPES: [&str; 17] = [
    "text",
    "image",
    "document",
    "tool_use",
    "tool_result",
    "thinking",
    "redacted_thinking",
    "server_tool_use",
    "web_search_tool_result",
    "web_fetch_tool_result",
    "code_execution_tool_result",
    "bash_code_execution_tool_result",
    "text_editor_code_execution_tool_result",
    "tool_search_tool_result",
    "search_result",
    "container_upload",
    "tool_reference",
];

/// TS `BlobRef` — `{ $blob: "sha256:<hex>", bytes }` (decoded byte length).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BlobRef {
    pub blob: String,
    pub bytes: i64,
}

impl BlobRef {
    /// The reference as JSON, keys in TS order (`$blob`, `bytes`).
    pub fn to_value(&self) -> Value {
        let mut map = Map::new();
        map.insert("$blob".into(), Value::String(self.blob.clone()));
        map.insert("bytes".into(), Value::Number(self.bytes.into()));
        Value::Object(map)
    }

    /// The bare hex digest (`sha256:` prefix removed when present).
    pub fn sha256_hex(&self) -> &str {
        self.blob.strip_prefix("sha256:").unwrap_or(&self.blob)
    }
}

/// TS `ExtractedBlob`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ExtractedBlob {
    pub sha256: String,
    pub media_type: Option<String>,
    pub data: Vec<u8>,
}

/// TS `ExtractResult`.
#[derive(Debug, Clone, PartialEq)]
pub struct ExtractResult {
    pub blocks: Vec<Value>,
    pub blobs: Vec<ExtractedBlob>,
}

/// TS `BlobLoader = (sha256) => Uint8Array | undefined`.
pub type BlobLoader<'a> = &'a dyn Fn(&str) -> Option<Vec<u8>>;

pub fn is_plain_record(value: &Value) -> bool {
    value.is_object()
}

pub fn is_api_block(value: &Value) -> bool {
    match value.as_object().and_then(|o| o.get("type")) {
        Some(Value::String(t)) => API_BLOCK_TYPES.contains(&t.as_str()),
        _ => false,
    }
}

pub fn is_blob_ref(value: &Value) -> bool {
    match value.as_object() {
        Some(o) => {
            matches!(o.get("$blob"), Some(Value::String(_)))
                && matches!(o.get("bytes"), Some(Value::Number(_)))
        }
        None => false,
    }
}

/// Parse a `{ $blob, bytes }` value (see [`is_blob_ref`]).
pub fn blob_ref_of(value: &Value) -> Option<BlobRef> {
    let o = value.as_object()?;
    let blob = o.get("$blob")?.as_str()?.to_string();
    let bytes = match o.get("bytes")? {
        Value::Number(n) => n
            .as_i64()
            .or_else(|| n.as_f64().map(|f| f as i64))
            .unwrap_or(0),
        _ => return None,
    };
    Some(BlobRef { blob, bytes })
}

fn blob_ref_for(data: &[u8]) -> (BlobRef, String) {
    let sha256 = sha256_hex_bytes(data);
    (
        BlobRef {
            blob: format!("sha256:{sha256}"),
            bytes: data.len() as i64,
        },
        sha256,
    )
}

// ── base64 (Node `Buffer` semantics) ─────────────────────────────

fn base64_value(c: u8) -> Option<u8> {
    match c {
        b'A'..=b'Z' => Some(c - b'A'),
        b'a'..=b'z' => Some(c - b'a' + 26),
        b'0'..=b'9' => Some(c - b'0' + 52),
        b'+' | b'-' => Some(62),
        b'/' | b'_' => Some(63),
        _ => None,
    }
}

/// `Buffer.from(text, "base64")`: lenient — both alphabets, characters
/// outside the alphabet skipped, padding optional, decoding stops at the
/// first `=`, a trailing partial group yields its whole bytes only.
pub fn base64_decode_node(text: &str) -> Vec<u8> {
    let mut out = Vec::with_capacity(text.len() * 3 / 4);
    let mut acc: u32 = 0;
    let mut bits: u32 = 0;
    for &c in text.as_bytes() {
        if c == b'=' {
            break;
        }
        let Some(v) = base64_value(c) else {
            continue;
        };
        acc = (acc << 6) | u32::from(v);
        bits += 6;
        if bits >= 8 {
            bits -= 8;
            out.push(((acc >> bits) & 0xff) as u8);
        }
    }
    out
}

/// `Buffer.from(bytes).toString("base64")`: standard alphabet, padded.
pub fn base64_encode(data: &[u8]) -> String {
    const ALPHABET: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    let mut out = String::with_capacity(data.len().div_ceil(3) * 4);
    for chunk in data.chunks(3) {
        let b0 = chunk[0];
        let b1 = chunk.get(1).copied().unwrap_or(0);
        let b2 = chunk.get(2).copied().unwrap_or(0);
        let n = (u32::from(b0) << 16) | (u32::from(b1) << 8) | u32::from(b2);
        out.push(ALPHABET[((n >> 18) & 63) as usize] as char);
        out.push(ALPHABET[((n >> 12) & 63) as usize] as char);
        if chunk.len() > 1 {
            out.push(ALPHABET[((n >> 6) & 63) as usize] as char);
        } else {
            out.push('=');
        }
        if chunk.len() > 2 {
            out.push(ALPHABET[(n & 63) as usize] as char);
        } else {
            out.push('=');
        }
    }
    out
}

// ── extraction (intake) ──────────────────────────────────────────

struct Sink {
    blobs: Vec<ExtractedBlob>,
}

/// A `{type:"base64", media_type, data}` source: data leaves as a blob. Other
/// source types (url, text, content) are text-shaped or nested.
fn extract_source(source: &Value, sink: &mut Sink) -> Value {
    let Some(src) = source.as_object() else {
        return source.clone();
    };
    if src.get("type").and_then(Value::as_str) == Some("base64")
        && let Some(Value::String(text)) = src.get("data")
    {
        let data = base64_decode_node(text);
        let (reference, sha256) = blob_ref_for(&data);
        sink.blobs.push(ExtractedBlob {
            sha256,
            media_type: src
                .get("media_type")
                .and_then(Value::as_str)
                .map(str::to_string),
            data,
        });
        let mut out = src.clone();
        out.insert("data".into(), reference.to_value());
        return Value::Object(out);
    }
    if src.get("type").and_then(Value::as_str) == Some("content")
        && let Some(Value::Array(items)) = src.get("content")
    {
        let mut out = src.clone();
        out.insert(
            "content".into(),
            Value::Array(
                items
                    .iter()
                    .map(|inner| extract_block(inner, sink))
                    .collect(),
            ),
        );
        return Value::Object(out);
    }
    source.clone()
}

fn extract_opaque(block: &Map<String, Value>, key: &str, sink: &mut Sink) -> Value {
    let Some(Value::String(text)) = block.get(key) else {
        return Value::Object(block.clone());
    };
    let data = text.as_bytes().to_vec();
    let (reference, sha256) = blob_ref_for(&data);
    sink.blobs.push(ExtractedBlob {
        sha256,
        media_type: None,
        data,
    });
    let mut out = block.clone();
    out.insert(key.to_string(), reference.to_value());
    Value::Object(out)
}

/// One block, blob payloads extracted, everything else verbatim. Unknown
/// shapes pass through untouched (a newer API type is still a record).
fn extract_block(value: &Value, sink: &mut Sink) -> Value {
    let Some(block) = value.as_object() else {
        return value.clone();
    };
    match block.get("type").and_then(Value::as_str) {
        Some("image") | Some("document") => {
            let mut out = block.clone();
            let source = block.get("source").cloned().unwrap_or(Value::Null);
            out.insert("source".into(), extract_source(&source, sink));
            Value::Object(out)
        }
        Some("redacted_thinking") => extract_opaque(block, "data", sink),
        Some("tool_result") => match block.get("content") {
            Some(Value::Array(items)) => {
                let mut out = block.clone();
                out.insert(
                    "content".into(),
                    Value::Array(
                        items
                            .iter()
                            .map(|inner| extract_block(inner, sink))
                            .collect(),
                    ),
                );
                Value::Object(out)
            }
            _ => value.clone(),
        },
        Some("web_search_tool_result") => match block.get("content") {
            Some(Value::Array(items)) => {
                let mut out = block.clone();
                out.insert(
                    "content".into(),
                    Value::Array(
                        items
                            .iter()
                            .map(|inner| match inner.as_object() {
                                Some(o)
                                    if o.get("type").and_then(Value::as_str)
                                        == Some("web_search_result") =>
                                {
                                    extract_opaque(o, "encrypted_content", sink)
                                }
                                _ => inner.clone(),
                            })
                            .collect(),
                    ),
                );
                Value::Object(out)
            }
            _ => value.clone(),
        },
        Some("web_fetch_tool_result") => {
            // content is a web_fetch block whose `content` is a document block.
            match block.get("content").and_then(Value::as_object) {
                Some(fetch) if fetch.get("content").is_some_and(Value::is_object) => {
                    let mut inner_out = fetch.clone();
                    let doc = fetch.get("content").cloned().unwrap_or(Value::Null);
                    inner_out.insert("content".into(), extract_block(&doc, sink));
                    let mut out = block.clone();
                    out.insert("content".into(), Value::Object(inner_out));
                    Value::Object(out)
                }
                _ => value.clone(),
            }
        }
        _ => value.clone(),
    }
}

/// Blocks with their binary/opaque payloads replaced by blob references.
pub fn extract_blobs(blocks: &[Value]) -> ExtractResult {
    let mut sink = Sink { blobs: Vec::new() };
    let out = blocks
        .iter()
        .map(|block| extract_block(block, &mut sink))
        .collect();
    ExtractResult {
        blocks: out,
        blobs: sink.blobs,
    }
}

// ── inlining (serving) ───────────────────────────────────────────

#[derive(Clone, Copy)]
enum InlineEncoding {
    Base64,
    Utf8,
}

fn inline_value(value: &Value, load: BlobLoader<'_>, encoding: InlineEncoding) -> Value {
    if let Some(reference) = blob_ref_of(value) {
        let Some(data) = load(reference.sha256_hex()) else {
            return value.clone();
        };
        return Value::String(match encoding {
            InlineEncoding::Base64 => base64_encode(&data),
            InlineEncoding::Utf8 => String::from_utf8_lossy(&data).into_owned(),
        });
    }
    value.clone()
}

/// The block as the API shaped it, blob payloads back in place. A blob that
/// is missing from the store leaves its reference in place rather than
/// inventing bytes.
pub fn inline_blobs(block: &Value, load: BlobLoader<'_>) -> Value {
    match block {
        Value::Array(items) => Value::Array(items.iter().map(|b| inline_blobs(b, load)).collect()),
        Value::Object(map) => {
            let mut out = Map::new();
            for (key, value) in map {
                if is_blob_ref(value) {
                    // base64 sources come back as base64; opaque strings (redacted thinking
                    // data, encrypted search content) were stored as their utf8 bytes.
                    let encoding = if key == "data"
                        && map.get("type").and_then(Value::as_str) == Some("base64")
                    {
                        InlineEncoding::Base64
                    } else {
                        InlineEncoding::Utf8
                    };
                    out.insert(key.clone(), inline_value(value, load, encoding));
                } else if value.is_object() || value.is_array() {
                    out.insert(key.clone(), inline_blobs(value, load));
                } else {
                    out.insert(key.clone(), value.clone());
                }
            }
            Value::Object(out)
        }
        other => other.clone(),
    }
}

// ── text-shaped rendering (bands, derivations, retrieval) ────────

/// JS `Number.prototype.toFixed(digits)` for the finite, non-negative values
/// this module formats: nearest, ties toward the larger value.
fn js_to_fixed(x: f64, digits: u32) -> String {
    let scale = 10f64.powi(digits as i32);
    let scaled = x * scale;
    let floor = scaled.floor();
    let n = if scaled - floor >= 0.5 {
        floor + 1.0
    } else {
        floor
    };
    if digits == 0 {
        return format!("{}", n as i64);
    }
    let whole = (n / scale).floor() as i64;
    let frac = (n as i64) - whole * (scale as i64);
    format!("{whole}.{frac:0width$}", width = digits as usize)
}

fn format_bytes(bytes: i64) -> String {
    if bytes < 1024 {
        return format!("{bytes} B");
    }
    if bytes < 1024 * 1024 {
        return format!("{} KB", js_to_fixed(bytes as f64 / 1024.0, 0));
    }
    format!("{} MB", js_to_fixed(bytes as f64 / (1024.0 * 1024.0), 1))
}

fn source_description(source: &Value) -> String {
    let Some(src) = source.as_object() else {
        return String::new();
    };
    let media_type = src.get("media_type").and_then(Value::as_str);
    let data = src.get("data");
    let source_type = src.get("type").and_then(Value::as_str);
    let joined = |size: String| -> String {
        // `[mediaType, size].filter(Boolean).join(" · ")`.
        let mut parts: Vec<String> = Vec::new();
        if let Some(mt) = media_type
            && !mt.is_empty()
        {
            parts.push(mt.to_string());
        }
        if !size.is_empty() {
            parts.push(size);
        }
        parts.join(" · ")
    };
    if source_type == Some("base64")
        && let Some(reference) = data.and_then(blob_ref_of)
    {
        return joined(format_bytes(reference.bytes));
    }
    if source_type == Some("base64")
        && let Some(Value::String(text)) = data
    {
        // JS string length is UTF-16 code units; base64 text is ASCII.
        let len = text.encode_utf16().count() as i64;
        return joined(format_bytes(len * 3 / 4));
    }
    if source_type == Some("url")
        && let Some(Value::String(url)) = src.get("url")
    {
        return url.clone();
    }
    if source_type == Some("text") {
        return "text/plain".to_string();
    }
    if source_type == Some("content") {
        return "content".to_string();
    }
    String::new()
}

/// What the model needs to know that the block existed, not its content: the
/// API type, media type, size, and a title when the block has one. Text-shaped
/// blocks render their text. Tool-result nesting renders each inner block.
pub fn placeholder_text(block: &Value) -> String {
    let Some(b) = block.as_object() else {
        return String::new();
    };
    let str_field = |key: &str| -> Option<&str> { b.get(key).and_then(Value::as_str) };
    match b.get("type").and_then(Value::as_str) {
        Some("text") => str_field("text").unwrap_or("").to_string(),
        Some("image") => format!(
            "[image · {}]",
            source_description(b.get("source").unwrap_or(&Value::Null))
        ),
        Some("document") => {
            let title = match str_field("title") {
                Some(t) if !t.is_empty() => format!(" · {t}"),
                _ => String::new(),
            };
            let source = b.get("source").unwrap_or(&Value::Null);
            if let Some(src) = source.as_object()
                && src.get("type").and_then(Value::as_str) == Some("text")
                && let Some(Value::String(data)) = src.get("data")
            {
                return data.clone();
            }
            format!("[document · {}{title}]", source_description(source))
        }
        Some("redacted_thinking") => "[redacted thinking]".to_string(),
        Some("search_result") => {
            let title = str_field("title").unwrap_or("");
            let source = str_field("source").unwrap_or("");
            let inner = match b.get("content") {
                Some(Value::Array(items)) => items
                    .iter()
                    .map(placeholder_text)
                    .collect::<Vec<_>>()
                    .join("\n"),
                _ => String::new(),
            };
            format!("[search result · {title} · {source}]\n{inner}")
        }
        Some("tool_reference") => format!(
            "[tool reference · {}]",
            js_string_nullish(b.get("tool_name"))
        ),
        Some("container_upload") => format!(
            "[container upload · {}]",
            js_string_nullish(b.get("file_id"))
        ),
        Some("server_tool_use") => {
            let input = match b.get("input") {
                Some(Value::Null) | None => Value::Object(Map::new()),
                Some(other) => other.clone(),
            };
            format!(
                "[server tool use · {}] {}",
                js_string_nullish(b.get("name")),
                js_json_stringify(&input)
            )
        }
        Some("web_search_tool_result") => match b.get("content") {
            Some(Value::Array(items)) => {
                let lines: Vec<String> = items
                    .iter()
                    .map(|r| match r.as_object() {
                        Some(o) => format!(
                            "- {} {}",
                            js_string_nullish(o.get("title")),
                            js_string_nullish(o.get("url"))
                        )
                        .trim()
                        .to_string(),
                        None => String::new(),
                    })
                    .collect();
                format!(
                    "[web search result · {} result(s)]\n{}",
                    items.len(),
                    lines.join("\n")
                )
            }
            other => {
                let content = match other {
                    Some(Value::Null) | None => Value::Object(Map::new()),
                    Some(v) => v.clone(),
                };
                format!("[web search result] {}", js_json_stringify(&content))
            }
        },
        Some("web_fetch_tool_result") => {
            let content = b.get("content").and_then(Value::as_object);
            let url = content
                .and_then(|c| c.get("url"))
                .and_then(Value::as_str)
                .unwrap_or("");
            let doc = match content.and_then(|c| c.get("content")) {
                Some(inner) => placeholder_text(inner),
                None => match content {
                    // `placeholderText(undefined)` → "".
                    Some(_) | None => String::new(),
                },
            };
            format!("[web fetch result · {url}]\n{doc}")
        }
        Some("tool_result") => match b.get("content") {
            Some(Value::String(s)) => s.clone(),
            Some(Value::Array(items)) => items
                .iter()
                .map(placeholder_text)
                .collect::<Vec<_>>()
                .join("\n"),
            _ => String::new(),
        },
        // code_execution_tool_result and friends are text-shaped JSON; anything
        // newer is at least made visible.
        _ => js_json_stringify(block),
    }
}

/// True when the block carries (or carried) a blob payload.
pub fn has_blob_payload(block: &Value) -> bool {
    match block {
        Value::Array(items) => items.iter().any(has_blob_payload),
        Value::Object(map) => map
            .values()
            .any(|value| is_blob_ref(value) || has_blob_payload(value)),
        _ => false,
    }
}

/// Rough context cost of a block the text estimator cannot see. Images: the
/// API's ceiling for one image (~1,600 tokens at 1568px); PDFs: ~2,000 tokens
/// per page at ~50 KB a page, floor one page. Measured, not calibrated.
pub fn blob_token_estimate(block: &Value) -> i64 {
    let Some(b) = block.as_object() else {
        return 0;
    };
    let bytes = b
        .get("source")
        .and_then(Value::as_object)
        .and_then(|src| src.get("data"))
        .and_then(blob_ref_of)
        .map(|r| r.bytes)
        .unwrap_or(0);
    match b.get("type").and_then(Value::as_str) {
        Some("image") => 1_600,
        Some("document") => {
            if bytes > 0 {
                // Math.max(1, Math.ceil(bytes / 50_000)) * 2_000
                std::cmp::max(1, (bytes + 49_999) / 50_000) * 2_000
            } else {
                0
            }
        }
        Some("tool_result") => match b.get("content") {
            Some(Value::Array(items)) => items.iter().map(blob_token_estimate).sum(),
            _ => 0,
        },
        Some("web_fetch_tool_result") => match b.get("content").and_then(Value::as_object) {
            Some(content) => content.get("content").map(blob_token_estimate).unwrap_or(0),
            None => 0,
        },
        _ => 0,
    }
}
